lib/Parse/ParseExpr.cpp

```diff
@@ -3023,6 +3022,74 @@ ExprResult Parser::ParseBoundsExpression() {
+    return Actions.getASTContext().getPointerType(newPointee, pType->getKind());
+  }
+  case Type::ConstantArray: {
+    const ConstantArrayType *caType = dyn_cast<ConstantArrayType>(T);
+    QualType EltTy = SubstituteTypeVariable(caType->getElementType(), typeNames);
+    const llvm::APInt ArySize = caType->getSize();
+    ArrayType::ArraySizeModifier ASM = caType->getSizeModifier();
+    unsigned IndexTypeQuals = caType->getIndexTypeCVRQualifiers();
+    bool isChecked = caType->isChecked();
+    return Actions.getASTContext().getConstantArrayType(EltTy, ArySize, ASM, 
+                                                  IndexTypeQuals, isChecked);
+  }
+  case Type::VariableArray: {
+    const VariableArrayType *vaType = dyn_cast<VariableArrayType>(T);
+    QualType EltTy = SubstituteTypeVariable(vaType->getElementType(), typeNames);
+    Expr *NumElts = vaType->getSizeExpr();
+    ArrayType::ArraySizeModifier ASM = vaType->getSizeModifier();
+    unsigned IndexTypeQuals = vaType->getIndexTypeCVRQualifiers();
+    SourceRange Brackets = vaType->getBracketsRange();
+    return Actions.getASTContext().getVariableArrayType(EltTy, NumElts, ASM, 
+                                                    IndexTypeQuals, Brackets);
+  }
+  case Type::IncompleteArray: {
+    const IncompleteArrayType *iaType = dyn_cast<IncompleteArrayType>(T);
+    QualType EltTy = SubstituteTypeVariable(iaType->getElementType(), typeNames);
+    ArrayType::ArraySizeModifier ASM = iaType->getSizeModifier();
+    unsigned IndexTypeQuals = iaType->getIndexTypeCVRQualifiers();
+    bool isChecked = iaType->isChecked();
+    return Actions.getASTContext().getIncompleteArrayType(EltTy, ASM, 
+                                                    IndexTypeQuals, isChecked);
+  }
+  case Type::DependentSizedArray: {
+    const DependentSizedArrayType *dsaType = dyn_cast<DependentSizedArrayType>(T);
+
+    QualType EltTy = SubstituteTypeVariable(dsaType->getElementType(), typeNames);
+    Expr *NumElts = dsaType->getSizeExpr();
+    ArrayType::ArraySizeModifier ASM = dsaType->getSizeModifier();
+    unsigned IndexTypeQuals = dsaType->getIndexTypeCVRQualifiers();
+    SourceRange Brackets = dsaType->getBracketsRange();
+    return Actions.getASTContext().getDependentSizedArrayType(EltTy, NumElts, 
+                                                ASM, IndexTypeQuals, Brackets);
+  }
+
+  case Type::Typedef: {
+    // If underlying type is TypeVariableType, must replace the entire
+    // TypedefType. If underlying type is not TypeVariableType, there cannot be
+    // a type variable type that this context can replace.
+    const TypedefType *tdType = dyn_cast<TypedefType>(T);
+    const Type *underlying = tdType->getDecl()->getUnderlyingType().getTypePtr();
+    if (const TypeVariableType *tvType = dyn_cast<TypeVariableType>(underlying))
+      return typeNames[tvType->GetIndex()].typeName;
+    else
+      return QT;
+  }
```

Comment:
It is possible for various transformations in clang to eliminate the typedef.  Please add a case for type variable by itself.    Also, please add a case for function types.  Function types could use a type variable or a type that contains a type variable.